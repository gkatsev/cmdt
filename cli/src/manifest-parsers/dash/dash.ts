import {
	getMediaTypeFromMimeType,
	type ImageRepresentation,
	type Manifest,
	type ManifestParser,
	MediaType,
	type Representation,
	type Segment,
	UniqueRepresentationMap,
} from "cmdt-shared";
import { deepmergeCustom } from "deepmerge-ts";
import type winston from "winston";
import { getInstance as getLogger } from "../../logger.js";
import getStreamAndLanguages from "../../utils/cea/getStreamAndLanguages.js";
import ECeaSchemeUri from "../../utils/manifest/enum/ECeaSchemeUri.js";
import { secondsToMilliseconds } from "../../utils/time-utils.js";
import { wrapUrl } from "../../utils/url.js";
import {
	type AdaptationSet,
	type ContentType,
	getRawDashManifest,
	type MPD,
	type Period,
	type Representation as RawRepresentation,
	type SegmentTemplate,
} from "./raw-dash.js";
export class DashManifest implements ManifestParser {
	private logger: winston.Logger;
	private manifest: Manifest;
	constructor() {
		this.logger = getLogger();
		this.manifest = {
			url: wrapUrl("http://localhost"), // Placeholder
			video: new UniqueRepresentationMap(),
			audio: new UniqueRepresentationMap(),
			images: new UniqueRepresentationMap(),
			captionStreamToLanguage: {},
		};
	}
	public async parse(manifest: string, manifestUrl: string): Promise<Manifest> {
		const mpd = await getRawDashManifest(manifest);
		this.manifest.url = new URL(manifestUrl);
		this.manifest = this.parseRawManifest(mpd);
		return this.manifest;
	}

	private parseRawManifest(mpd: MPD): Manifest {
		for (const period of mpd.periods) {
			this.processPeriod(period);
		}

		return this.manifest;
	}

	private contentTypeToMediaType(contentType: ContentType): MediaType {
		switch (contentType) {
			case "video":
				return MediaType.Video;
			case "audio":
				return MediaType.Audio;
			case "image":
				return MediaType.Image;
			case "text":
				return MediaType.Text;
			case "application":
			case "font":
				return MediaType.Video;
		}
	}

	private guessMediaTypeFromCodecs(codecs: string): MediaType {
		if (codecs.includes("avc1") || codecs.includes("hev1") || codecs.includes("hvc1")) {
			return MediaType.Video;
		}
		if (codecs.includes("mp4a") || codecs.includes("ac-3") || codecs.includes("ec-3")) {
			return MediaType.Audio;
		}
		return MediaType.Video;
	}

	private getMediaTypeFromRepresentation(representation: RawRepresentation): MediaType {
		let mediaType: MediaType;
		const codecs = representation.codecs ?? representation.adaptationSet.codecs;
		const mimeType = representation.mimeType ?? representation.adaptationSet.mimeType;
		if (mimeType) {
			mediaType = getMediaTypeFromMimeType(mimeType);
		} else if (representation.adaptationSet.contentType) {
			mediaType = this.contentTypeToMediaType(representation.adaptationSet.contentType);
		} else if (codecs) {
			mediaType = this.guessMediaTypeFromCodecs(codecs);
		} else {
			this.logger.warn(`Could not determine media type for representation ${representation.id}`);
			mediaType = MediaType.Unknown;
		}
		return mediaType;
	}

	private buildSegmentUrlFromTemplate(
		baseUrl: string,
		segmentNumber: number,
		representation: RawRepresentation,
		uriTemplate: string,
	): string {
		const widthStr = uriTemplate.match(/\$Number%?0?([0-9]*)d?\$/)?.[1];
		const width = widthStr ? Number.parseInt(widthStr, 10) : 0;
		const paddedNumber = segmentNumber.toString().padStart(width, "0");
		if (baseUrl.endsWith("/")) {
			baseUrl = baseUrl.slice(0, -1);
		}
		let url = uriTemplate.startsWith("http")
			? uriTemplate.replace(/\$Number%?0?[0-9]*d?\$/, paddedNumber)
			: `${baseUrl}/${uriTemplate.replace(/\$Number%?0?[0-9]*d?\$/, paddedNumber)}`;
		url = url.replace(/\$RepresentationID\$/, representation.id);
		return url;
	}

	private getBaseUrl(representation: RawRepresentation): string {
		let baseUrls: Array<string> = [];

		if (representation.adaptationSet.period.manifest.baseUrl?.[0]) {
			baseUrls.push(representation.adaptationSet.period.manifest.baseUrl[0].url);
		}

		if (representation.adaptationSet.period.baseUrl?.[0]) {
			baseUrls.push(representation.adaptationSet.period.baseUrl[0].url);
		}

		if (representation.adaptationSet.baseUrl?.[0]) {
			baseUrls.push(representation.adaptationSet.baseUrl[0].url);
		}

		if (representation.baseUrl?.[0]) {
			baseUrls.push(representation.baseUrl[0].url);
		}

		const firstAbsolute = baseUrls.findIndex((url) => url.startsWith("http"));

		if (firstAbsolute >= 0) {
			baseUrls = baseUrls.slice(firstAbsolute);
		} else {
			baseUrls = [this.manifest.url.href, ...baseUrls];
		}

		let absoluteBase = new URL(this.manifest.url.href);
		for (const baseUrl of baseUrls) {
			if (baseUrl.startsWith("http")) {
				absoluteBase = new URL(baseUrl);
			} else {
				absoluteBase = new URL(baseUrl, absoluteBase);
			}
		}
		return absoluteBase.href;
	}

	private getSegmentsFromSegmentTemplate(
		representation: RawRepresentation,
		segmentTemplate: SegmentTemplate,
	): Array<Segment> {
		const segments: Array<Segment> = [];

		let n = segmentTemplate.startNumber ?? 1;
		const periodStart = representation.adaptationSet.period.start ?? 0;
		const baseUrl = this.getBaseUrl(representation);
		const timescale = segmentTemplate.timescale ?? 1;

		if (!segmentTemplate.media) {
			this.logger.warn(`No media template for representation ${representation.id}`);
			return [];
		}

		for (const timeline of segmentTemplate.segmentTimeline?.s ?? []) {
			const numSegments = timeline.r + 1;
			const tWithOffset = (timeline.t ?? 0) - (segmentTemplate.presentationTimeOffset ?? 0);
			const unscaledDuration = timeline.d ?? 0;
			for (let i = 0; i < numSegments; i++) {
				segments.push({
					initSegmentUrl: segmentTemplate.initialization
						? this.buildSegmentUrlFromTemplate(baseUrl, n, representation, segmentTemplate.initialization)
						: undefined,
					duration: secondsToMilliseconds(unscaledDuration / timescale),
					startTime: secondsToMilliseconds(periodStart + (tWithOffset + i * unscaledDuration) / timescale),
					url: this.buildSegmentUrlFromTemplate(baseUrl, n, representation, segmentTemplate.media),
					rawSegmentTime: secondsToMilliseconds(((timeline.t ?? 0) + i * unscaledDuration) / timescale),
				});
				n++;
			}
		}

		return segments;
	}

	private getSegmentsFromRepresentation(representation: RawRepresentation): Array<Segment> {
		if (representation.segmentTemplate || representation.adaptationSet.segmentTemplate) {
			const mergedTemplate = deepmergeCustom({ mergeArrays: false })(
				representation.adaptationSet.segmentTemplate ?? {},
				representation.segmentTemplate,
			) as SegmentTemplate;
			return this.getSegmentsFromSegmentTemplate(representation, mergedTemplate);
		}
		this.logger.warn(`No segment template for representation ${representation.id}`);
		return [];
	}

	private parseVideoRepresentation(representation: RawRepresentation): void {
		const hasCea608 = representation.adaptationSet.accessibility?.some((e) => e.schemeIdUri === ECeaSchemeUri.CEA608);
		const hasCea708 = representation.adaptationSet.accessibility?.some((e) => e.schemeIdUri === ECeaSchemeUri.CEA708);
		const videoRepresentation: Representation = {
			id: representation.id,
			width: representation.width ?? representation.adaptationSet.width,
			height: representation.height ?? representation.adaptationSet.height,
			bandwidth: representation.bandwidth,
			type: MediaType.Video,
			hasCaptions: {
				cea608: hasCea608 ?? false,
				cea708: hasCea708 ?? false,
			},
			codecs: representation.codecs ?? representation.adaptationSet.codecs,
			language: representation.adaptationSet.lang,
			segments: this.getSegmentsFromRepresentation(representation),
		};
		if (hasCea608 || hasCea708) {
			if (!representation.adaptationSet.accessibility) {
				throw new Error(`No accessibility information found for adaptation set ${representation.adaptationSet.id}`);
			}
			for (const accessibility of representation.adaptationSet.accessibility) {
				const info = getStreamAndLanguages(accessibility);
				for (const entry of info) {
					this.manifest.captionStreamToLanguage[entry[0]] = entry[1];
				}
			}
		}
		this.manifest.video.add(videoRepresentation);
	}

	private parseAudioRepresentation(representation: RawRepresentation): void {
		const audioRepresentation: Representation = {
			id: representation.id,
			width: representation.width ?? representation.adaptationSet.width,
			height: representation.height ?? representation.adaptationSet.height,
			bandwidth: representation.bandwidth,
			type: MediaType.Audio,
			hasCaptions: {
				cea608: false,
				cea708: false,
			},
			codecs: representation.codecs ?? representation.adaptationSet.codecs,
			language: representation.adaptationSet.lang,
			segments: this.getSegmentsFromRepresentation(representation),
		};
		this.manifest.audio.add(audioRepresentation);
	}

	private parseImageRepresentation(representation: RawRepresentation): void {
		let dashThumbProperty = representation.essentialProperty?.find(
			(e) => e.schemeIdUri === "http://dashif.org/guidelines/thumbnail_tile",
		);
		if (!dashThumbProperty) {
			dashThumbProperty = representation.adaptationSet.essentialProperty?.find(
				(e) => e.schemeIdUri === "http://dashif.org/guidelines/thumbnail_tile",
			);
		}

		if (!dashThumbProperty?.value) {
			this.logger.warn(`No thumbnail tile information found for representation ${representation.id}`);
			return;
		}

		const [cols, rows] = dashThumbProperty.value.split("x").map((val) => Number.parseInt(val, 10));

		if (!cols || !rows) {
			this.logger.warn(`Invalid thumbnail tile information found for representation ${representation.id}`);
			return;
		}

		const imageRepresentation: ImageRepresentation = {
			id: representation.id,
			width: representation.width ?? representation.adaptationSet.width,
			height: representation.height ?? representation.adaptationSet.height,
			bandwidth: representation.bandwidth,
			type: MediaType.Image,
			hasCaptions: {
				cea608: false,
				cea708: false,
			},
			codecs: representation.codecs ?? representation.adaptationSet.codecs,
			language: representation.adaptationSet.lang,
			imageRows: rows,
			imageCols: cols,
			segments: this.getSegmentsFromRepresentation(representation),
		};
		this.manifest.images.add(imageRepresentation);
	}

	private parseRepresentation(representation: RawRepresentation): void {
		const mediaType: MediaType = this.getMediaTypeFromRepresentation(representation);
		switch (mediaType) {
			case MediaType.Video:
				this.parseVideoRepresentation(representation);
				break;
			case MediaType.Audio:
				this.parseAudioRepresentation(representation);
				break;
			case MediaType.Image:
				this.parseImageRepresentation(representation);
				break;
			// case MediaType.Text:
			// 	this.parseTextRepresentation(representation);
			// 	break;
		}
	}

	private processAdaptationSet(adaptationSet: AdaptationSet): void {
		for (const representation of adaptationSet.representation ?? []) {
			this.parseRepresentation(representation);
		}
	}

	private processPeriod(period: Period): void {
		for (const adaptationSet of period.adaptationSet ?? []) {
			this.processAdaptationSet(adaptationSet);
		}
	}
}
