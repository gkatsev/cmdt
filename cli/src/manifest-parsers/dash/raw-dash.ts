import { parse as parseDuration, toSeconds } from "iso8601-duration";
import xml2js from "xml2js";
import { parseBooleans, parseNumbers } from "xml2js/lib/processors.js";

// biome-ignore lint/suspicious/noExplicitAny: XML Node type isn't known during build time
export type XmlNode = any;

type Seconds = number;

type BaseUrl = {
	url: string;
	serviceLocation?: string;
	byteRange?: string;
	availabilityTimeOffset?: string;
	availabilityTimeComplete?: string;
	timeShiftBufferDepth?: Seconds;
	rangeAccess?: boolean;
}

type Location = {

}

type ServiceDescription = {
	
}

type ProgramInformation = {

}

type EventStream = {

}

export type Url = {
	sourceURL?: string;
	range?: string;
}

export type SegmentTimeline = {
	s?: Array<{
		t?: number;
		n?: number;
		d?: number;
		r: number;
	}>;
};

export type MultipleSegmentBaseInformation = {
	duration?: number;
	startNumber?: number;
	endNumber?: number;
	segmentTimeline?: SegmentTimeline;

} & SegmentBase;

export type SegmentBase = {
 timescale: number;
 presentationTimeOffset?: number;
 presentationDuration?: number;
 timeShiftBufferDepth?: Seconds;
 indexRange?: string;
availabilityTimeOffset?: number;
	availabilityTimeComplete?: boolean;
	initializationElement?: Url;
	representationIndex?: Url;
}

type SegmentUrl = {
	media?: string;
	mediaRange?: string;
	index?: string;
	indexRange?: string;
}

export type SegmentList = {
	segmentUrl?: SegmentUrl[];
} & MultipleSegmentBaseInformation;

export type SegmentTemplate = {
	media?: string;
	index?: string;
	initialization?: string;
	bitstreamSwitching?: string;
} & MultipleSegmentBaseInformation;

export type Descriptor = {
	schemeIdUri: string;
	value?: string;
	id?: string;
}

type Label = {
	id?: string;
	lang?: string;
	value: string;

}

type ProducerReferenceTime = {
	id: string;
	inBand?: boolean;
	type?: string;
	applicationScheme?: string;
	wallClockTime: string;
	presentationTime: string;
	utcTiming?: Descriptor;
}

type RepresentationBase = {
	profiles?: string;
	width?: number;
	height?: number;
	sar?: string;
	frameRate?: string;
	audioSamplingRate?: string;
	mimeType?: string;
	segmentProfiles?: string;
	codecs?: string;
	containerProfiles?: string;
	maximumSAPPeriod?: number;
	startWithSAP?: string;
	maxPlayoutRate?: number;
	codingDependency?: boolean;
	scanType?: string;
	selectionPriority?: number;
	tag?: string;
	framePacking?: Descriptor[];
	audioChannelConfiguration?: Descriptor[];
	contentProtection?: ContentProtection[];
	outputProtection?: Descriptor[];
	essentialProperty?: Descriptor[];
	supplementalProperty?: Descriptor[];
	inbandEventStream?: EventStream[];
	label?: Label[];
	groupLabel?: Label[];
	producerReferenceTime?: ProducerReferenceTime[];
}

export type ContentType = 'video' | 'audio' | 'image' | 'text' | 'application' | 'font';

export type Representation = {
	id: string;
	bandwidth: number;
	baseUrl?: BaseUrl[];
	segmentBase?: SegmentBase;
	segmentList?: SegmentList;
	segmentTemplate?: SegmentTemplate;
    adaptationSet: AdaptationSet;
} & RepresentationBase;

export type AdaptationSet = {
	id?: string;
	lang?: string;
	contentType?: ContentType;
	par?: string;
	minBandwidth?: number;
	maxBandwidth?: number;
	minWidth?: number;
	maxWidth?: number;
	minHeight?: number;
	maxHeight?: number;
	minFrameRate?: string;
	maxFrameRate?: string;
	segmentAlignment?: boolean;
	bitstreamSwitching?: boolean;
	subsegmentAlignment?: boolean;
	subsegmentStartsWithSAP?: number;
	accessibility?: Descriptor[];
	role?: Descriptor[];
	baseUrl?: BaseUrl[];
	segmentBase?: SegmentBase;
	segmentList?: SegmentList;
	segmentTemplate?: SegmentTemplate;
	representation?: Representation[];
    period: Period;

} & RepresentationBase;

export type Period = {
	id?: string;
	start?: Seconds;
	bitstreamSwitching?: boolean;
	baseUrl?: BaseUrl[];
	segmentBase?: SegmentBase;
	segmentList?: SegmentList;
	segmentTemplate?: SegmentTemplate;
	eventStream?: EventStream[];
	contentProtection?: ContentProtection[];
	adaptationSet?: AdaptationSet[];
    manifest: MPD;
}

export type ContentProtection = {
    schemeIdUri: string;
    value?: string;
    ref?: string;
    refId?: string;
    robustness?: string;
};

export type MPD = {
	id?: string;
	profiles: string;
	type: 'static' | 'dynamic'; // Default 'static'
	availabilityStartTime?: string; // Required for dynamic
	publishTime?: string; // Required for dynamic
	availabilityEndTime?: string;
	mediaPresentationDuration?: Seconds;
	minimumUpdatePeriod?: Seconds;
	minBufferTime: Seconds;
	timeShiftBufferDepth?: Seconds;
	suggestedPresentationDelay?: Seconds;
	maxSegmentDuration?: Seconds;
	maxSubsegmentDuration?: Seconds;
	baseUrl?: BaseUrl[];
	location?: Location[];
	serviceDescription?: ServiceDescription[];
	programInformation?: ProgramInformation[];
	periods: Period[];
	contentProtection?: ContentProtection[];
}


function optionalDurationToSeconds(input?: string): Seconds | undefined {
	if (!input) {
		return undefined;
	}
	return toSeconds(parseDuration(input));
}

function requiredDurationToSeconds(input: string): Seconds {
    return toSeconds(parseDuration(input));
}

function parseBaseUrls(baseUrlRoot: XmlNode): BaseUrl[] | undefined {
    return baseUrlRoot?.map((e: XmlNode) => {
        return {
            url: e._,
            serviceLocation: e.$?.serviceLocation,
            byteRange: e.$?.byteRange,
            availabilityTimeOffset: e.$?.availabilityTimeOffset,
            availabilityTimeComplete: e.$?.availabilityTimeComplete,
            timeShiftBufferDepth: optionalDurationToSeconds(e.$?.timeShiftBufferDepth),
            rangeAccess: e.$?.rangeAccess,
        }
    });
}

function parseContentProtection(contentProtectionRoot: XmlNode): ContentProtection[] | undefined {
    return contentProtectionRoot?.map((e: XmlNode) => {
        return {
            schemeIdUri: e.$.schemeIdUri,
            value: e.$?.value,
            ref: e.$?.ref,
            refId: e.$?.refId,
            robustness: e.$?.robustness,
        }
    });
}

function parseUrl(urlRoot: XmlNode): Url | undefined {
    if(!urlRoot) {
        return undefined;
    }
    return {
        sourceURL: urlRoot.$.sourceURL,
        range: urlRoot.$.range,
    }
}

function parseSegmentBase(segmentBaseRoot: XmlNode): SegmentBase | undefined {
    if(!segmentBaseRoot) {
        return undefined;
    }

    return {
        timescale: segmentBaseRoot.$.timescale,
        presentationTimeOffset: segmentBaseRoot.$.presentationTimeOffset,
        presentationDuration: segmentBaseRoot.$.presentationDuration,
        timeShiftBufferDepth: optionalDurationToSeconds(segmentBaseRoot.$.timeShiftBufferDepth),
        indexRange: segmentBaseRoot.$.indexRange,
        availabilityTimeOffset: segmentBaseRoot.$.availabilityTimeOffset,
        availabilityTimeComplete: segmentBaseRoot.$.availabilityTimeComplete,
        initializationElement: parseUrl(segmentBaseRoot.Initialization?.[0]),
        representationIndex: parseUrl(segmentBaseRoot.RepresentationIndex?.[0]),
    }
}

function parseSegmentTimeline(segmentTimelineRoot: XmlNode): SegmentTimeline | undefined {
    if(!segmentTimelineRoot) {
        return undefined;
    }

    return {
        s: segmentTimelineRoot.S?.map((e: XmlNode) => {
            return {
                t: e.$.t,
                n: e.$.n,
                d: e.$.d,
                r: e.$.r ?? 0,
            }
        }),
    }
}

function parseSegmentUrl(segmentUrlRoot: XmlNode): SegmentUrl[] | undefined {
    return segmentUrlRoot?.map((e: XmlNode) => {
        return {
            media: e.$.media,
            mediaRange: e.$.mediaRange,
            index: e.$.index,
            indexRange: e.$.indexRange,
        }
    });
}

function parseDescriptor(descriptorRoot: XmlNode): Descriptor[] | undefined {
    return descriptorRoot?.map((e: XmlNode) => {
        return {
            schemeIdUri: e.$.schemeIdUri,
            value: e.$.value,
            id: e.$.id,
        }
    });
}

function parseLabel(labelRoot: XmlNode): Label[] | undefined {
    return labelRoot?.map((e: XmlNode) => {
        return {
            id: e.$.id,
            lang: e.$.lang,
            value: e._,
        }
    });
}

function parseProducerReferenceTime(producerReferenceTimeRoot: XmlNode): ProducerReferenceTime[] | undefined {
    return producerReferenceTimeRoot?.map((e: XmlNode) => {
        return {
            id: e.$.id,
            inBand: e.$.inBand,
            type: e.$.type,
            applicationScheme: e.$.applicationScheme,
            wallClockTime: e.$.wallClockTime,
            presentationTime: e.$.presentationTime,
            utcTiming: parseDescriptor(e.UTCTiming),
        }
    });
}

function parseRepresentationBase(representationBaseRoot: XmlNode): RepresentationBase | undefined {
    if(!representationBaseRoot) {
        return undefined;
    }

    return {
        profiles: representationBaseRoot.$.profiles,
        width: representationBaseRoot.$.width,
        height: representationBaseRoot.$.height,
        sar: representationBaseRoot.$.sar,
        frameRate: representationBaseRoot.$.frameRate,
        audioSamplingRate: representationBaseRoot.$.audioSamplingRate,
        mimeType: representationBaseRoot.$.mimeType,
        segmentProfiles: representationBaseRoot.$.segmentProfiles,
        codecs: representationBaseRoot.$.codecs,
        containerProfiles: representationBaseRoot.$.containerProfiles,
        maximumSAPPeriod: representationBaseRoot.$.maximumSAPPeriod,
        startWithSAP: representationBaseRoot.$.startWithSAP,
        maxPlayoutRate: representationBaseRoot.$.maxPlayoutRate,
        codingDependency: representationBaseRoot.$.codingDependency,
        scanType: representationBaseRoot.$.scanType,
        selectionPriority: representationBaseRoot.$.selectionPriority,
        tag: representationBaseRoot.$.tag,
        framePacking: parseDescriptor(representationBaseRoot.FramePacking),
        audioChannelConfiguration: parseDescriptor(representationBaseRoot.AudioChannelConfiguration),
        contentProtection: parseContentProtection(representationBaseRoot.ContentProtection),
        outputProtection: parseDescriptor(representationBaseRoot.OutputProtection),
        essentialProperty: parseDescriptor(representationBaseRoot.EssentialProperty),
        supplementalProperty: parseDescriptor(representationBaseRoot.SupplementalProperty),
        // inbandEventStream: parseEventStream(representationBaseRoot.InbandEventStream),
        label: parseLabel(representationBaseRoot.Label),
        groupLabel: parseLabel(representationBaseRoot.GroupLabel),
        producerReferenceTime: parseProducerReferenceTime(representationBaseRoot.ProducerReferenceTime),
    }
}

function parseMultiSegmentBaseInformation(multiSegmentBaseInformationRoot: XmlNode): MultipleSegmentBaseInformation {
    return {
        duration: multiSegmentBaseInformationRoot.$.duration,
        startNumber: multiSegmentBaseInformationRoot.$.startNumber,
        endNumber: multiSegmentBaseInformationRoot.$.endNumber,
        segmentTimeline: parseSegmentTimeline(multiSegmentBaseInformationRoot.SegmentTimeline?.[0]),
        initializationElement: parseUrl(multiSegmentBaseInformationRoot.Initialization?.[0]),
        indexRange: multiSegmentBaseInformationRoot.$.indexRange,
        availabilityTimeOffset: multiSegmentBaseInformationRoot.$.availabilityTimeOffset,
        availabilityTimeComplete: multiSegmentBaseInformationRoot.$.availabilityTimeComplete,
        representationIndex: parseUrl(multiSegmentBaseInformationRoot.RepresentationIndex?.[0]),
        timescale: multiSegmentBaseInformationRoot.$.timescale,
        presentationTimeOffset: multiSegmentBaseInformationRoot.$.presentationTimeOffset,
        presentationDuration: multiSegmentBaseInformationRoot.$.presentationDuration,
        timeShiftBufferDepth: optionalDurationToSeconds(multiSegmentBaseInformationRoot.$.timeShiftBufferDepth),
        
    }
}

function parseSegmentTemplate(segmentTemplateRoot: XmlNode): SegmentTemplate | undefined {
    if(!segmentTemplateRoot) {
        return undefined;
    }
    const multiSegmentInfo = parseMultiSegmentBaseInformation(segmentTemplateRoot);
    return {
        media: segmentTemplateRoot.$.media,
        index: segmentTemplateRoot.$.index,
        initialization: segmentTemplateRoot.$.initialization,
        bitstreamSwitching: segmentTemplateRoot.$.bitstreamSwitching,
        ...multiSegmentInfo,

    }
}

function parseSegmentList(segmentListRoot: XmlNode): SegmentList | undefined {
    if(!segmentListRoot) {
        return undefined;
    }

    return {
        duration: segmentListRoot.$.duration,
        startNumber: segmentListRoot.$.startNumber,
        endNumber: segmentListRoot.$.endNumber,
        segmentTimeline: parseSegmentTimeline(segmentListRoot.SegmentTimeline?.[0]),
        initializationElement: parseUrl(segmentListRoot.Initialization?.[0]),
        segmentUrl: parseSegmentUrl(segmentListRoot.SegmentURL),
        timescale: segmentListRoot.$.timescale,
        presentationTimeOffset: segmentListRoot.$.presentationTimeOffset,
        presentationDuration: segmentListRoot.$.presentationDuration,
        timeShiftBufferDepth: optionalDurationToSeconds(segmentListRoot.$.timeShiftBufferDepth),
        indexRange: segmentListRoot.$.indexRange,
        availabilityTimeOffset: segmentListRoot.$.availabilityTimeOffset,
        availabilityTimeComplete: segmentListRoot.$.availabilityTimeComplete,
        representationIndex: parseUrl(segmentListRoot.RepresentationIndex?.[0]),
    }
}

function parseRepresentation(representationRoot: XmlNode, adaptationSet: AdaptationSet): Representation[] | undefined {
    return representationRoot?.map((e: XmlNode) => {
        return {
            id: e.$.id,
            bandwidth: e.$.bandwidth,
            baseUrl: parseBaseUrls(e.BaseURL),
            segmentBase: parseSegmentBase(e.SegmentBase?.[0]),
            segmentList: parseSegmentList(e.SegmentList?.[0]),
            segmentTemplate: parseSegmentTemplate(e.SegmentTemplate?.[0]),
            adaptationSet,
            ...parseRepresentationBase(e),
        }
    });
}

function parseAdaptationSet(adaptationSetRoot: XmlNode, period: Period): AdaptationSet[] | undefined {
    return adaptationSetRoot?.map((e: XmlNode) => {
        const adaptationSet: AdaptationSet = {
            id: e.$?.id,
            lang: e.$?.lang,
            contentType: e.$?.contentType,
            par: e.$?.par,
            minBandwidth: e.$?.minBandwidth,
            maxBandwidth: e.$?.maxBandwidth,
            minWidth: e.$?.minWidth,
            maxWidth: e.$?.maxWidth,
            minHeight: e.$?.minHeight,
            maxHeight: e.$?.maxHeight,
            minFrameRate: e.$?.minFrameRate,
            maxFrameRate: e.$?.maxFrameRate,
            segmentAlignment: e.$?.segmentAlignment,
            bitstreamSwitching: e.$?.bitstreamSwitching,
            subsegmentAlignment: e.$?.subsegmentAlignment,
            subsegmentStartsWithSAP: e.$?.subsegmentStartsWithSAP,
            accessibility: parseDescriptor(e.Accessibility),
            role: parseDescriptor(e.Role),
            baseUrl: parseBaseUrls(e.BaseURL),
            segmentBase: parseSegmentBase(e.SegmentBase?.[0]),
            segmentList: parseSegmentList(e.SegmentList?.[0]),
            segmentTemplate: parseSegmentTemplate(e.SegmentTemplate?.[0]),
            ...parseRepresentationBase(e),
            period,
        }
        adaptationSet.representation = parseRepresentation(e.Representation, adaptationSet);
        return adaptationSet;
    });
}

export function parsePeriods(periodRoot: XmlNode, mpd: MPD): Period[] {
    return periodRoot.map((e: XmlNode) => {
        const period: Period = {
            id: e.$?.id,
            start: optionalDurationToSeconds(e.$?.start),
            bitstreamSwitching: e.$?.bitstreamSwitching,
            baseUrl: parseBaseUrls(e.BaseURL),
            segmentBase: parseSegmentBase(e.SegmentBase?.[0]),
            segmentList: parseSegmentList(e.SegmentList?.[0]),
            segmentTemplate: parseSegmentTemplate(e.SegmentTemplate?.[0]),
            // eventStream: parseEventStream(e.EventStream),
            contentProtection: parseContentProtection(e.ContentProtection),
            manifest: mpd,
        }
        period.adaptationSet = parseAdaptationSet(e.AdaptationSet, period);
        return period;
    });
}

export async function getRawDashManifest(manifest: string, manifestUrl: string): Promise<MPD> {
    const raw = await xml2js.parseStringPromise(manifest, { explicitArray: true, explicitCharkey: true, attrValueProcessors: [parseNumbers, parseBooleans] });
    const mpd: MPD = {
        profiles: raw.MPD.$.profiles,
        id: raw.MPD.$.id,
        type: raw.MPD.$.type ?? 'static',
        availabilityStartTime: raw.MPD.$.availabilityStartTime,
        publishTime: raw.MPD.$.publishTime,
        availabilityEndTime: raw.MPD.$.availabilityEndTime,
        mediaPresentationDuration: optionalDurationToSeconds(raw.MPD.$.mediaPresentationDuration),
        minimumUpdatePeriod: optionalDurationToSeconds(raw.MPD.$.minimumUpdatePeriod),
        minBufferTime: requiredDurationToSeconds(raw.MPD.$.minBufferTime)!,
        timeShiftBufferDepth: optionalDurationToSeconds(raw.MPD.$.timeShiftBufferDepth),
        suggestedPresentationDelay: optionalDurationToSeconds(raw.MPD.$.suggestedPresentationDelay),
        maxSegmentDuration: optionalDurationToSeconds(raw.MPD.$.maxSegmentDuration),
        maxSubsegmentDuration: optionalDurationToSeconds(raw.MPD.$.maxSubsegmentDuration),
        baseUrl: parseBaseUrls(raw.MPD.BaseURL),
        periods: [],
        contentProtection: parseContentProtection(raw.MPD.ContentProtection),
    }
    mpd.periods = parsePeriods(raw.MPD.Period, mpd);
    return mpd;
}