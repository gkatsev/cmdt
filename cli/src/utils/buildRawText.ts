import type { Text } from "cmdt-shared";

const buildRawText = (texts: Array<Text>): string => {
	return texts
		.map((t: Text) => t.text.trim())
		.join(" ")
		.replace(/[\n\r\t]|\s+/gm, " ")
		.trim();
};

export default buildRawText;
