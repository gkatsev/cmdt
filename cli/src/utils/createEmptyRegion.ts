import type { Region } from "cmdt-shared";

const createEmptyRegion = (): Region => {
	return {
		width: 100,
		height: 16,
		regionAnchorX: 0,
		regionAnchorY: 0,
		viewportanchorX: 0,
		viewportanchorY: 0,
		align: "center",
		scroll: "up",
		style: null,
	};
};

export default createEmptyRegion;
