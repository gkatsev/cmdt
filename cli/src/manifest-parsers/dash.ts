import {
	getMediaTypeFromMimeType,
	type ImageRepresentation,
	type Manifest,
	type ManifestParser,
	MediaType,
	type Representation,
	type Segment,
} from "cmdt-shared";
import { getRawDashManifest } from "./raw-dash.js";
export class DashManifest implements ManifestParser {
	public async parse(manifest: string, manifestUrl: string): Promise<Manifest> {
		const mpd = await getRawDashManifest(manifest, manifestUrl);
		return this.parseRawManifest(mpd, manifestUrl);
	}
}
