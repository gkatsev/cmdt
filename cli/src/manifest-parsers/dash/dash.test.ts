import { parse } from "@fast-csv/parse";
import axios from "axios";
import fs from "fs";
import { beforeAll, describe, expect, it } from "vitest";
import { getTestFile, writeTestFile } from "../../../test/utils.js";
import { DashManifest } from "./dash.js";

describe("DashManifest", () => {
	it("should parse a DASH manifest", async () => {
		const manifestUrl = "http://example.com/manifest.mpd";
		const testManifest = await getTestFile("manifests/dash-multiperiod.mpd");
		const parser = new DashManifest();
		const manifest = await parser.parse(testManifest, manifestUrl);
		console.log(JSON.stringify(manifest, null, 2));
		expect(manifest).toMatchSnapshot();
	});
	// describe("DASH-IF test vectors", async () => {
	// 	const dashTestVectors = await getTestFile("manifests/DASH IF Test Assets Database.csv");
	// 	await new Promise<void>((resolve, reject) => {
	// 		const stream = parse({ delimiter: ";", headers: true })
	// 			.on("error", (error) => reject(error))
	// 			.on("data", (row) => {
	// 				it(`should parse vector ${row["Testvector"]}`, async () => {
	// 					// Check to see if manifest is cached
	// 					let manifest: string;
	// 					const manifestURL = new URL(row["URL"]);
	// 					const manifestFileName = manifestURL.pathname.split("/").pop();

	// 					try {
	// 						manifest = await getTestFile(`manifests/dash-if-cache/${manifestFileName}`);
	// 					} catch (e) {
	// 						manifest = (await axios(row["URL"])).data;
	// 						await writeTestFile(`manifests/dash-if-cache/${manifestFileName}`, manifest);
	// 					}

	// 					const parser = new DashManifest();
	// 					const parsedManifest = await parser.parse(manifest, row["URL"]);
	// 					expect(parsedManifest).toBeDefined();
	// 				});
	// 			})
	// 			.on("end", (rowCount: number) => resolve());
	// 		stream.write(dashTestVectors);
	// 		stream.end();
	// 	});
	// });
});
