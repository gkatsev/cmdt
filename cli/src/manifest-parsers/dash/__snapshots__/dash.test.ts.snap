// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DashManifest > should parse a DASH manifest 1`] = `
{
  "audio": [],
  "captionStreamToLanguage": {},
  "images": [],
  "url": "http://example.com/manifest.mpd",
  "video": [],
}
`;
